### ISSUE TITLE
some layers size change to zero

### ISSUE THREAD

**Issue by xuejipeng** [docker/distribution#2788]
when i pull a image，registry get  a erro 'swift: Timeout expired while waiting for segments of /docker/registry/v2/blobs/sha256/74/74a09ea38d47837fc3b0e1d209b9ce035d242db0c882ded3fdb6861a1fb558e2/data to show up"',  and i saw the data size is zero, what actions will cause the size of this file turn into 0?

![image](https://user-images.githubusercontent.com/18901031/50069505-5d01c580-0205-11e9-8ee1-283d65e7c951.png)

**Comment by ZiggyMaes** [comment]
Related to https://github.com/docker/distribution/issues/2758.

**Comment by andyzhangx** [comment]
I hit the same issue (layer size is 0) on disk file system pulling `library/nginx`, and I am using `registry:2`, restart the container would always work, below are the error logs:
```
# wget  --no-check-certificate https://127.0.0.1:5000/v2/library/nginx/blobs/sha256:177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be
--2019-01-05 04:28:18--  https://127.0.0.1:5000/v2/library/nginx/blobs/sha256:177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be
Connecting to 127.0.0.1:5000... connected.
    WARNING: certificate common name ‘*.azk8s.cn’ doesn't match requested host name ‘127.0.0.1’.
HTTP request sent, awaiting response... 200 OK
Length: 22492916 (21M) [application/octet-stream]
Saving to: ‘sha256:177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be.3’

sha256:177e7ef0df6987e0c5738a1fb5aba98b   0%[                                                                             ]       0  --.-KB/s    in 0s

2019-01-05 04:28:18 (0.00 B/s) - Connection closed at byte 0. Retrying.
```

I could see it's returning 0 size in `registry:2` container logs:
```
{"log":"172.17.0.1 - - [05/Jan/2019:04:22:56 +0000] \"GET /v2/library/nginx/blobs/sha256:177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be HTTP/1.1\" 200 0 \"\" \"Wget/1.19.4 (linux-gnu)\"\n","stream":"stdout","time":"2019-01-05T04:22:56.038140961Z"}
{"log":"time=\"2019-01-05T04:22:58Z\" level=info msg=\"response completed\" go.version=go1.7.6 http.request.host=\"127.0.0.1:5000\" http.request.id=c0ae0353-f7fa-4756-8d8c-3a6581530230 http.request.method=GET http.request.remoteaddr=\"172.17.0.1:49052\" http.request.uri=\"/v2/library/nginx/blobs/sha256:177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be\" http.request.useragent=\"Wget/1.19.4 (linux-gnu)\" http.response.contenttype=\"application/octet-stream\" http.response.duration=3.750332ms http.response.status=200 http.response.written=0 instance.id=97ada566-7d80-468a-a8e9-beae2fe154e4 version=v2.6.2 \n","stream":"stdout","time":"2019-01-05T04:22:58.048987967Z"}
{"log":"172.17.0.1 - - [05/Jan/2019:04:22:58 +0000] \"GET 
```

I already hit this issue several times in two months, every time if I restart the registry container, it would work well. Looks like there is metadata corruption inside `registry:2` container.

**Comment by xuejipeng** [comment]
Are you using the Swift interface? is a issue for swift api. i make a shell scripts to check the zero layer and delete them.

**Comment by andyzhangx** [comment]
@xuejipeng No, I am using filesystem interface, I could see that sometimes the layer does not exist on local file system, and it return zero bytes directly instead of fetching from remote server.

**Comment by manishtomar** [comment]
@andyxning How was `library/nginx` uploaded to filesystem? Did you do `docker push 127.0.0.1:5000/library/nginx` or something else?

**Comment by andyzhangx** [comment]
@manishtomar Yes, my container registry acts as proxy, I won't do `docker push` by myself since it's a registry proxy.
actually I found the specific layer `177e7ef0df6987e0c5738a1fb5aba98b6b6e7a5fef992e481977dbb5ba3f91be` does not exist on my filesystem, while other layers of `library/nginx` is on my file system, I suspect there is metadata inconsistency inside registry logic, it occurs about only **once** a month, if I restart the registry, it will work again. Anyway, I have switched to use `registry:2.7.0`, not sure whether that version has this issue.
>@andyxning How was library/nginx uploaded to filesystem? Did you do docker push >127.0.0.1:5000/library/nginx or something else?

**Comment by andyzhangx** [comment]
@manishtomar another clue is that this issue could possibly happen when the original nginx image expires per my observation.

**Comment by manishtomar** [comment]
@andyzhangx You are most likely hitting #2367. Please confirm if that is the case and I can close this issue. Unfortunately there is no plan to fix that anytime soon.

**Comment by andyzhangx** [comment]
@manishtomar thanks for guiding me to the correct issue, so there is no one investigating at this issue?

**Comment by manishtomar** [comment]
At this time no one is looking into it.

**Comment by danielefranceschi** [comment]
Configured as a dockerhub mirror, the bug is still present in `registry:2.7.1`
```
10.9.8.7 - - [03/Dec/2021:13:32:00 +0000] "GET /v2/library/alpine/blobs/sha256:59bf1c3509f33515622619af21ed55bbe26d24913cedbca106468a5fb37a50c3?ns=docker.io HTTP/1.1" 200 0 "" "containerd/v1.4.3-k3s3"
```
Occurs after 2/3 days of uptime.

Restarting the mirror fixes the issue.
